addons/binding/org.openhab.binding.homie/src/main/java/org/openhab/binding/homie/internal/HomieChannelTypeProviderImpl.java

```diff
@@ -0,0 +1,139 @@
+    public ChannelGroupType getChannelGroupType(ChannelGroupTypeUID channelGroupTypeUID, Locale locale) {
+        return groups.get(channelGroupTypeUID);
+    }
+
+    @Override
+    public Collection<ChannelGroupType> getChannelGroupTypes(Locale locale) {
+        return groups.values();
+    }
+
+    @Override
+    public void addChannelType(ChannelTypeUID uid, boolean readOnly) {
+
+        String description = "";
+        String itemType = "String";
+        URI configDescriptionURI = null;
+        ChannelKind kind = ChannelKind.STATE;
+        BigDecimal state_min = null;
+        BigDecimal state_max = null;
+        BigDecimal state_step = null;
+        String state_pattern = null;
+        boolean state_readOnly = readOnly;
+        List<StateOption> state_options = Collections.emptyList();
+        StateDescription state = new StateDescription(state_min, state_max, state_step, state_pattern, state_readOnly,
+                state_options);
+        Set<String> tags = Collections.emptySet();
+        String label = uid.getId();
+        String category = "";
+        boolean advanced = false;
+
+        EventDescription event = null;
+        ChannelType type = new ChannelType(uid, advanced, itemType, kind, label, description, category, tags, state,
+                event, configDescriptionURI);
+        types.put(uid, type);
+    }
+
+    @Override
+    public void addChannelGroupType(ChannelGroupTypeUID uid, String label) {
+        String description = "";
+        boolean advanced = false;
+        List<ChannelDefinition> channelDefinitions = Collections.emptyList();
+        ChannelGroupType type = new ChannelGroupType(uid, advanced, label, description, channelDefinitions);
+        groups.put(uid, type);
+
+    }
+
+    @Override
+    public void addChannelToGroup(ChannelUID channelId, ChannelTypeUID channelTypeUid,
+            ChannelGroupTypeUID channelGroupId) {
+        ChannelGroupType chGrp = getChannelGroupType(channelGroupId, null);
+        ChannelDefinition chDef = new ChannelDefinition(channelId.getId(), channelTypeUid);
+        List<ChannelDefinition> channelDefinitions = new LinkedList<>(chGrp.getChannelDefinitions());
+        channelDefinitions.add(chDef);
+        ChannelGroupType newGroup = new ChannelGroupType(chGrp.getUID(), chGrp.isAdvanced(), chGrp.getLabel(),
+                chGrp.getDescription(), channelDefinitions);
+        groups.put(channelGroupId, newGroup);
+
+    }
+
+    @Override
+    public ChannelTypeUID createChannelTypeBySettings(String unit, BigDecimal min, BigDecimal max, BigDecimal step,
+            String itemType, boolean isReadonly, String category) {
+        String settingsIdentifier = String.format("gen-", UUID.randomUUID());
+
+        ChannelTypeUID uid = new ChannelTypeUID(BINDING_ID, settingsIdentifier);
+        String description = "";
+        URI configDescriptionURI = null;
+        ChannelKind kind = ChannelKind.STATE;
+        BigDecimal state_min = min;
```

Comment:
1) naming conventions
2) maybe this code is better of by not assigning them (`min`, `max`, `step`, `isReadonly`) to another fields since they are already properly named.